### PR TITLE
chore: upgrade TypeScript & Eslint & Typedoc

### PR DESCRIPTION
**Pull request by jbroma**
### Summary

Old versions of `typescript` are blocking fastify upgrade, there it's time to update. To support `typescript@5.x`, `eslint-confg-callstack` needed to be updated as well. `Eslint` and `TypeDoc` were upgraded as they became incompatible as well.

- [x] - upgrade `typescript` in all packages
- [x] - upgrade `eslint` in all packages
- [x] - upgrade `eslint-config-callstack`  in all packages
- [x] - upgrade `typedoc` & `typedoc-markdown-plugin` in dev-server & repack packages
- [x] - fix TypeScript issues
- [x] - fix Eslint issues

### Test plan

- [x] - tests pass
- [x] - lint pass
- [x] - typescript pass

### CODE REVIEW

File: packages/repack/src/webpack/Compiler.ts [hunk 1]
```diff
@@ -184,7 +184,7 @@ export class Compiler extends EventEmitter {
       this.progressSenders[platform] = this.progressSenders[platform] ?? [];
       this.progressSenders[platform].push(sendProgress);
     }
-    return new Promise<Asset>((resolve, reject) => {
+    return await new Promise<Asset>((resolve, reject) => {
```
Comment (jbroma) [inline packages/repack/src/webpack/Compiler.ts:187]: this is just here to satisfy the linter, has no effect on the result

File: packages/repack/src/modules/WebpackHMRClient.ts [hunk 2]
```diff
@@ -128,7 +128,8 @@ class HMRClient {
       });
 
       if (!this.upToDate()) {
-        this.checkUpdates(update);
+        // TODO - should we return here?
+        void this.checkUpdates(update);
```
Comment (jbroma) [inline packages/repack/src/modules/WebpackHMRClient.ts:132]: this should probably be handled in separate PR - as this would cause a change in behaviour, needs to be checked whether returning here would still result in correct behaviour

File: packages/dev-server/.eslintrc.cjs [hunk 3]
```diff
@@ -1,5 +1,14 @@
...
   extends: '@callstack/eslint-config/node',
+  parserOptions: {
+    project: true,
+    tsconfigRootDir: __dirname,
+  },
+  rules: {
+    /* These 2 rules do not play nicely with fastify */
+    'require-await': 0,
+    '@typescript-eslint/no-floating-promises': 0,
+  },
```
Comment (jbroma) [inline packages/dev-server/.eslintrc.cjs:11]: as the comment mentions, lot's of Fastify API uses `thenables` which incorrectly get picked up as `Promise`, that's why these 2 rules were disabled

File: packages/repack/src/commands/start.ts [hunk 4]
```diff
@@ -172,16 +172,18 @@ export async function start(_: string[], config: Config, args: StartArguments) {
...
+                ([name, asset]) => ({
+                  name,
+                  size: asset.info.size,
+                })
+              )
             ),
-          getCompilationStats: async (platform) =>
-            compiler.statsCache[platform] ?? null,
+          getCompilationStats: (platform) =>
+            Promise.resolve(compiler.statsCache[platform] ?? null),
```
Comment (jbroma) [inline packages/repack/src/commands/start.ts:186]: changed `async` to `Promise.resolve` to satisfy the requirements, otherwise TS & eslint have a conflict

**Review by jbroma — COMMENTED** [review]
Highlights

**Review by thymikee — APPROVED** [review]
